pkg/server/secrets-encrypt.go

```diff
@@ -329,26 +344,100 @@ func encryptionRotateKeys(ctx context.Context, server *config.Control) error {
+		err := retry.RetryOnConflict(retry.DefaultRetry, func() error {
+			node, err := server.Runtime.Core.Core().V1().Node().Get(nodeName, metav1.GetOptions{})
+			if err != nil {
+				return err
+			}
+			secretsencrypt.BootstrapEncryptionHashAnnotation(node, server.Runtime)
+			_, err = server.Runtime.Core.Core().V1().Node().Update(node)
+			return err
+		})
+		return err
+	}
+
+	// Remove last key
+	curKeys, err := secretsencrypt.GetEncryptionKeys(server.Runtime, false)
 	if err != nil {
 		return err
 	}
 
-	reencryptHash, err := secretsencrypt.GenReencryptHash(server.Runtime, secretsencrypt.EncryptionReencryptRequest)
+	logrus.Infoln("Removing key: ", curKeys[len(curKeys)-1])
+	curKeys = curKeys[:len(curKeys)-1]
+	if err = secretsencrypt.WriteEncryptionConfig(server.Runtime, curKeys, true); err != nil {
+		return err
+	}
+
+	if err = retry.RetryOnConflict(retry.DefaultRetry, func() error {
+		node, err := server.Runtime.Core.Core().V1().Node().Get(nodeName, metav1.GetOptions{})
+		if err != nil {
+			return err
+		}
+		return secretsencrypt.WriteEncryptionHashAnnotation(server.Runtime, node, false, secretsencrypt.EncryptionReencryptFinished)
+	}); err != nil {
+		return err
+	}
+
+	return cluster.Save(ctx, server, true)
+}
+
+func updateSecrets(ctx context.Context, server *config.Control, nodeName string) error {
+	restConfig, err := clientcmd.BuildConfigFromFlags("", server.Runtime.KubeConfigSupervisor)
+	if err != nil {
+		return err
+	}
+	// For secrets we need a much higher QPS than default
+	restConfig.QPS = secretsencrypt.SecretQPS
+	restConfig.Burst = secretsencrypt.SecretBurst
+	k8s, err := kubernetes.NewForConfig(restConfig)
 	if err != nil {
 		return err
 	}
-	ann := secretsencrypt.EncryptionReencryptRequest + "-" + reencryptHash
-	node.Annotations[secretsencrypt.EncryptionHashAnnotation] = ann
-	if _, err = server.Runtime.Core.Core().V1().Node().Update(node); err != nil {
+
+	nodeRef := &corev1.ObjectReference{
+		Kind:      "Node",
+		Name:      nodeName,
+		UID:       types.UID(nodeName),
+		Namespace: "",
+	}
+
+	// For backwards compatibility with the old controller, we use an event recorder instead of logrus
+	recorder := util.BuildControllerEventRecorder(k8s, "secrets-reencrypt", metav1.NamespaceDefault)
+
+	secretPager := pager.New(pager.SimplePageFunc(func(opts metav1.ListOptions) (runtime.Object, error) {
+		return k8s.CoreV1().Secrets(metav1.NamespaceAll).List(ctx, opts)
+	}))
+	secretPager.PageSize = secretsencrypt.SecretListPageSize
+
+	i := 0
+	if err := secretPager.EachListItem(ctx, metav1.ListOptions{}, func(obj runtime.Object) error {
+		secret, ok := obj.(*corev1.Secret)
+		if !ok {
+			return errors.New("failed to convert object to Secret")
+		}
+		if _, err := k8s.CoreV1().Secrets(secret.Namespace).Update(ctx, secret, metav1.UpdateOptions{}); err != nil && !apierrors.IsConflict(err) {
+			recorder.Eventf(secret, corev1.EventTypeNormal, secretsencrypt.SecretsUpdateErrorEvent, "failed to update secret: %v", err)
```

Comment:
nit: I think this can be a logged warning instead of an event. If you do want to keep it as an event:
1. attach the event to the node instead of the secret, otherwise it won't show up alongside the rest of the secret encrypt events on the node.
2. make it a Warning instead of a Normal event.